lib/cog/adapters/config.ex

```diff
@@ -43,154 +49,215 @@ defmodule Cog.Adapters.Config do
 
-      def start_link() do
-        GenServer.start_link(__MODULE__, [], name: __MODULE__)
+      def fetch_config do
+        Cog.Adapters.Config.fetch_config(unquote(key), unquote(schema))
       end
 
-      def init(_) do
-        {:ok, walk_config}
+      def fetch_config! do
+        Cog.Adapters.Config.fetch_config!(unquote(key), unquote(schema))
       end
+    end
+  end
 
-      def handle_call(:get_config, _from, state) do
-        {:reply, state, state}
-      end
+  # TODO: Alter the schema structure to deterministically identify namespaces
+  # from two tuple value definitions.
+  def fetch_config(key, schema) do
+    config = Application.get_env(:cog, key)
 
-      def walk_config(), do: walk_config(@schema)
-      def walk_config(schema) do
-        case Keyword.keyword?(schema) do
-          true ->
-            for key <- Keyword.keys(schema), into: %{},
-              do: {key, process_entries(Keyword.get(schema, key))}
-          false ->
-            process_entries(schema)
-        end
-      end
+    case Keyword.keyword?(schema) do
+      true ->
+        namespace_config(schema, config)
+      false ->
+        apply_schema(schema, config)
+    end
+  end
 
-      defp process_entries(entries) do
-        for entry <- entries, into: %{}, do: process_entry(entry)
-      end
+  def fetch_config!(key, schema) do
+    case fetch_config(key, schema) do
+      {:ok, config} ->
+        config
+      {:error, errors} ->
+        Enum.each(format_errors(errors), &Logger.error/1)
+        raise "Invalid config found for #{inspect(key)}"
+    end
+  end
 
-      # Field with no options and no source mapping.
-      defp process_entry(field) when is_atom(field) do
-        process_entry({field, [], field})
+  # The schema is namespaced by keys to seperate out the config values.
+  # Traverse one level down and run that schema on the top-level config.
+  #
+  # TODO: Support any level of namespacing by adding some fancy recursion.
+  defp namespace_config(schema, config) do
+    Enum.reduce(schema, {:ok, %{}}, fn {namespace, schema}, {state, acc} ->
+      case apply_schema(schema, config) do
+        {^state, value} ->
+          {state, Map.put(acc, namespace, value)}
+        {:error, value} ->
+          {:error, Map.put(%{}, namespace, value)}
+        _ ->
+          {state, acc}
       end
+    end)
+  end
 
-      # Hardcoded field.
-      defp process_entry({field, :hardcode, value}) do
-        {field, value}
+  # Iterate through the schema, find matching keys from the config, and run the
+  # set of rules in the order defined against the value output from the
+  # previous rule.
+  defp apply_schema(schema, config) do
+    Enum.reduce(schema, {:ok, %{}}, fn field, {state, acc} ->
+      case fetch_field(config, field) do
+        {^state, {key, value}} ->
+          {state, Map.put(acc, key, value)}
+        {:error, {key, errors}} ->
+          {:error, Map.put(%{}, key, errors)}
+        _ ->
+          {state, acc}
```

Comment:
After more thought, this still seemed like the most understandable solution. Mapping and then partitioning is pretty messy because we also have keys and nil values to worry about. But if anyone wants to pair on writing a more clear solution here, I'm up for it.